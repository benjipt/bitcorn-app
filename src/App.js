import React, { Component } from 'react'
import SignInPage from './components/SignInPage'
import AppBar from './components/AppBar'
import BalanceCard from './components/BalanceCard'
import SendCard from './components/SendCard'

const baseURL = 'https://jobcoin.gemini.com/greyhound-abruptly/api/'

export default class App extends Component {
  constructor(props) {
    super(props)

    this.state = {
      isLoggedIn: false,
      loggedInAddress: '',
      balance: '',
      transactions: [],
      runningBalance: []
    }

    this.getData = this.getData.bind(this)
    this.setRunningBalance = this.setRunningBalance.bind(this)
    this.handleLogout = this.handleLogout.bind(this)
  }

  getData(address) {
    fetch(baseURL + 'addresses/' + address)
      .then(data => { return data.json()}, err => console.log(err))
      .then(parsedData => {
        this.setState({ 
        isLoggedIn: true,
        loggedInAddress: address,
        balance: parsedData.balance,
        transactions: parsedData.transactions })
        this.setRunningBalance(this.state.transactions)
      }, err => console.log(err))
  }

  setRunningBalance(transactions) {
    let currentBalance = 0
    for (let transaction of transactions) {
      if (transaction.toAddress === this.state.loggedInAddress) {
        currentBalance += Number(transaction.amount)
      }
    }
    console.log(currentBalance)
  }

  handleLogout() {
    this.setState({
      isLoggedIn: false,
      loggedInAddress: '',
      balance: '',
      transactions: []
    })
  }

  render() {
    return (
      <div>
        { !this.state.isLoggedIn && 
          <SignInPage
            getData={ this.getData } /> }

        { this.state.isLoggedIn &&
          <AppBar 
            handleLogout={ this.handleLogout }
            address={ this.state.loggedInAddress } /> }

        { this.state.isLoggedIn &&
          <BalanceCard 
            balance={ this.state.balance } /> }

        { this.state.isLoggedIn &&
          <SendCard 
            address={ this.state.loggedInAddress }
            getData={ this.getData } /> }
      </div>
    )
  }
}

