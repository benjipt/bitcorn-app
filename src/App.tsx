import { useState, useEffect } from 'react'
import { format } from 'date-fns'
import SignInPage from './components/SignInPage'
import JobCoinUI from './components/JobCoinUI'
import { BalancePlot, Response, Transaction } from './types'

const baseURL = 'https://jobcoin.gemini.com/greyhound-abruptly/api/'

export default function App() {
  // STATE HOOKS
  const [ isLoggedIn, setIsLoggedIn ] = useState(false)
  const [ loggedInAddress, setLoggedInAddress ] = useState('')
  const [ balance, setBalance ] = useState('')
  const [ runningBalance, setRunningBalance ] = useState<BalancePlot[]>([])

  // PERSISTS STATE OF APPLICATION UPON REFRESH
  useEffect(() => {
    const persistedLoggedInAddress = sessionStorage.getItem('loggedInAddress')
    if (persistedLoggedInAddress && persistedLoggedInAddress !== 'null') {
      getData(persistedLoggedInAddress)
    }
    // eslint-disable-next-line
  }, [])

  // RETRIEVES ADDRESS DATA FROM API
  const getData = (address: string | null) => {
    if (address === null) return
    fetch(baseURL + 'addresses/' + address)
      .then(data => { return data.json()}, err => console.log(err))
      .then((parsedData: Response) => {
        const { balance, transactions } = parsedData
        setIsLoggedIn(true)
        setLoggedInAddress(address)
        setBalance(balance)
        createRunningBalance(transactions, address)
        // DUPLICATES ADDRESS TO LOCAL STORAGE FOR PERSISTENT STATE UPON REFRESH
        sessionStorage.setItem('loggedInAddress', address)
      }, err => console.log(err))
  }
  // SHAPES DATA FOR BALANCE HISTORY CHART ~~~~~>
  const createRunningBalance = (transactions: Transaction[], loggedInAddress: string) => {
    let currentBalance = 0
    let balanceArr: BalancePlot[] = []
    for (let transaction of transactions) {
      if (transaction.toAddress === loggedInAddress) {
        currentBalance += Number(transaction.amount)
      } else {
        currentBalance -= Number(transaction.amount)
      }
      const plot = generatePlot(transaction, currentBalance)
      updateBalanceArr(plot, balanceArr)
    }
    setRunningBalance(balanceArr)
  }
  // Formats dates for chart rendering
  const generatePlot = (transaction: Transaction, currentBalance: number) => {
    const formattedDate = format(new Date(transaction.timestamp), 'yyyy-MM-dd')
    return { amount: currentBalance, date: formattedDate }
  }
  // Ensures balance value from each day is latest balance
  const updateBalanceArr = (plot: BalancePlot, balanceArr: BalancePlot[]) => {
    const lastIndex = balanceArr.length - 1
    if (lastIndex > -1) {
      if (plot.date === balanceArr[lastIndex].date) {
        balanceArr.splice(lastIndex, 1, plot)
      } else {
        balanceArr.push(plot)
      }
    } else {
      balanceArr.push(plot)
    }
  }
  // <~~~~~ SHAPES DATA FOR BALANCE HISTORY CHART

  const handleLogout = () => {
    setIsLoggedIn(false)
    setLoggedInAddress('')
    setBalance('')
    setRunningBalance([])
    sessionStorage.clear()
  }

  return (
    <div data-testid="App-1">
      { !isLoggedIn && 
        <SignInPage
          getData={ getData } /> }

      { isLoggedIn &&
        <JobCoinUI 
          handleLogout={ handleLogout }
          address={ loggedInAddress }
          balance={ balance }
          getData={ getData }
          data={ runningBalance } /> }
    </div>
  )
}

