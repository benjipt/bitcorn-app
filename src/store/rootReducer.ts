// src/store/rootReducer.ts
import { combineReducers } from '@reduxjs/toolkit';
import userReducer from './slices/userSlice';
import uiReducer from './slices/uiSlice';

const rootReducer = combineReducers({
  user: userReducer,
  ui: uiReducer,
});

export default rootReducer;
