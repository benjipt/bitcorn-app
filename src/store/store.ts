// src/store/store.ts
import { configureStore } from '@reduxjs/toolkit';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import rootReducer from './rootReducer';

const store = configureStore({
  reducer: rootReducer,
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;

export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
export const useAppDispatch = () => useDispatch<AppDispatch>();

export default store;
