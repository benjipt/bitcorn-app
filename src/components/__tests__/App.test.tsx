import { describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '@/store/store';
import App from '@/App';

describe('App', () => {
  it('should render App', () => {
    // Renders App wrapped with the provider
    render(
      <Provider store={store}>
        <App />
      </Provider>
    );
    // Grabs App from render
    const AppComponent = screen.getByTestId('App-1');
    // Tests for expected App
    expect(AppComponent).toBeInTheDocument();
  });
});
