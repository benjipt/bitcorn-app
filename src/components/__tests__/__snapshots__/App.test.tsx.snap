// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > matches snapshot 1`] = `
<div
  data-testid="App-1"
>
  <div
    className="container text-center mt-5"
  >
    <h1>
      BITCORN
    </h1>
    <div
      className="mt-5 p-4 border-black rounded sign-in container-sm"
    >
      <p>
        Welcome! Please sign in with your Bitcorn address
      </p>
      <hr />
      <form
        className="mt-4"
        onSubmit={[Function]}
      >
        <div
          className="row g-3 align-items-center justify-content-center"
        >
          <div
            className="col-auto"
          >
            <label
              className="form-label visually-hidden"
              htmlFor="address"
            >
              Address
            </label>
          </div>
          <div
            className="col-auto"
          >
            <input
              className="form-control"
              id="address"
              onChange={[Function]}
              placeholder="Jilly"
              type="text"
              value=""
            />
          </div>
          <div
            className="col-auto"
          >
            <button
              className="btn btn-dark"
              data-testid="sign-in-btn"
              type="submit"
            >
              SIGN IN
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
